cranelift/codegen/src/legalizer/heap.rs

```diff
@@ -104,26 +113,27 @@ fn static_addr(
     isa: &dyn TargetIsa,
     inst: ir::Inst,
     heap: ir::Heap,
-    mut offset: ir::Value,
-    access_size: u64,
+    index: ir::Value,
+    offset: u32,
+    access_size: u8,
     bound: u64,
     func: &mut ir::Function,
     cfg: &mut ControlFlowGraph,
 ) {
-    let offset_ty = func.dfg.value_type(offset);
+    let index_ty = func.dfg.value_type(index);
     let addr_ty = func.dfg.value_type(func.dfg.first_result(inst));
     let mut pos = FuncCursor::new(func).at_inst(inst);
     pos.use_srcloc(inst);
 
-    // The goal here is to trap if `offset + access_size > bound`.
+    // The goal here is to trap if `index + offset + access_size > bound`.
     //
-    // This first case is a trivial case where we can easily trap.
-    if access_size > bound {
+    // This first case is a trivial case where we can statically trap.
+    if offset as u64 + access_size as u64 > bound {
```

Comment:
Seeing this a third time, now I think a helper might be best: `offset_plus_size(offset, access_size)`, giving us a nice typesafe clearly-overflow-safe idiom.